File: android/src/main/java/flutter/curiosity/scan/ScanView.java
package flutter.curiosity.scan;

import android.content.Context;
import android.graphics.ImageFormat;
import android.os.Build;
import android.util.Log;
import android.util.Size;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.camera.camera2.Camera2Config;
import androidx.camera.core.Camera;
import androidx.camera.core.CameraControl;
import androidx.camera.core.CameraInfo;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.CameraXConfig;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;
import androidx.camera.core.Preview;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.camera.view.PreviewView;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LifecycleRegistry;

import com.google.common.util.concurrent.ListenableFuture;
import com.google.zxing.BinaryBitmap;
import com.google.zxing.MultiFormatReader;
import com.google.zxing.PlanarYUVLuminanceSource;
import com.google.zxing.Result;
import com.google.zxing.common.HybridBinarizer;

import java.nio.ByteBuffer;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;

import flutter.curiosity.CuriosityPlugin;
import flutter.curiosity.utils.NativeUtils;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.platform.PlatformView;


public class ScanView implements PlatformView, LifecycleOwner, CameraXConfig.Provider,
        EventChannel.StreamHandler,
        MethodChannel.MethodCallHandler {

    private LifecycleRegistry lifecycleRegistry;
    private PreviewView previewView;
    private boolean isPlay;
    private EventChannel.EventSink eventSink;
    private long lastCurrentTimestamp = 0L;//最后一次的扫描

    private ListenableFuture<ProcessCameraProvider> cameraProviderFuture;
    private ProcessCameraProvider cameraProvider;
    private CameraControl cameraControl;
    private CameraInfo cameraInfo;
    private Context context;

    ScanView(Context ctx, BinaryMessenger messenger, int i, Object object) {
        context = ctx;
        Map map = (Map) object;
        isPlay = (Boolean) map.get("isPlay");
        int width = (int) map.get("width");
        int height = (int) map.get("height");
        new EventChannel(messenger, CuriosityPlugin.scanView + "_" + i + "/event")
                .setStreamHandler(this);
        MethodChannel methodChannel = new MethodChannel(messenger, CuriosityPlugin.scanView + "_" + i + "/method");
        methodChannel.setMethodCallHandler(this);
        previewView = initPreviewView(width, height);
        previewView.post(() -> startCamera(context, initPreview(width, height), initImageAnalysis(width, height)));
    }


    private PreviewView initPreviewView(int width, int height) {
        lifecycleRegistry = new LifecycleRegistry(this);
        cameraProviderFuture = ProcessCameraProvider.getInstance(context);
        previewView = new PreviewView(context);
        previewView.setLayoutParams(new ViewGroup.LayoutParams(width, height));
        previewView.setImplementationMode(PreviewView.ImplementationMode.TEXTURE_VIEW);
        return previewView;
    }


    private ImageAnalysis initImageAnalysis(int width, int height) {
        //设置分析
        ImageAnalysis imageAnalysis = new ImageAnalysis.Builder()
                .setImageQueueDepth(ImageAnalysis.STRATEGY_BLOCK_PRODUCER)
                .setTargetResolution(new Size(width, height))
                .build();
        imageAnalysis.setAnalyzer(Executors.newSingleThreadExecutor(), new ScanImageAnalysis());
        return imageAnalysis;
    }

    private Preview initPreview(int width, int height) {
        Preview preview = new Preview.Builder()
                .setTargetResolution(new Size(width, height))
                .build();
        preview.setSurfaceProvider(previewView.getPreviewSurfaceProvider());
        return preview;
    }

    private void startCamera(Context context, Preview preview, ImageAnalysis imageAnalysis) {
        CameraSelector cameraSelector = new CameraSelector.Builder().requireLensFacing(CameraSelector.LENS_FACING_BACK).build();
        cameraProviderFuture.addListener(() -> {
            try {
                cameraProvider = cameraProviderFuture.get();
                Camera camera = cameraProvider.bindToLifecycle(this, cameraSelector, preview,
                        imageAnalysis);
                cameraControl = camera.getCameraControl();
                cameraInfo = camera.getCameraInfo();
            } catch (ExecutionException | InterruptedException e) {
                e.printStackTrace();
            }
        }, ContextCompat.getMainExecutor(context));
    }

    @NonNull
    @Override
    public CameraXConfig getCameraXConfig() {
        return Camera2Config.defaultConfig();
    }


    private class ScanImageAnalysis implements ImageAnalysis.Analyzer {
        private MultiFormatReader multiFormatReader = new MultiFormatReader();

        @Override
        public void analyze(@NonNull ImageProxy image) {
            long currentTimestamp = System.currentTimeMillis();
            if (currentTimestamp - lastCurrentTimestamp >= 1L && isPlay == Boolean.TRUE) {
                if (ImageFormat.YUV_420_888 != image.getFormat()) {
                    return;
                }
                ByteBuffer buffer = image.getPlanes()[0].getBuffer();
                byte[] array = new byte[buffer.remaining()];

                buffer.get(array, 0, array.length);
                int height = image.getHeight();
                int width = image.getWidth();
                PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(array,
                        width,
                        height,
                        0,
                        0,
                        width,
                        height,
                        false);
                BinaryBitmap binaryBitmap = new BinaryBitmap(new HybridBinarizer(source.invert()));
//                multiFormatReader.setHints(NativeUtils.getHints());
                final Result result;
                try {
                    result = multiFormatReader.decode(binaryBitmap, NativeUtils.getHints());
//                    Log.i("扫码出来的数据", result.getText());
                    Log.i("扫码出来的数据", "");
                    if (result != null && eventSink != null) {
                        previewView.post(() -> {
                            if (eventSink != null)
                                eventSink.success(NativeUtils.scanDataToMap(result));
                        });
                    }
                } catch (Exception e) {
                    Log.i("无二维码数据", "无二维码数据");
                    buffer.clear();
                }
                lastCurrentTimestamp = currentTimestamp;
            }
            image.close();
        }
    }

    @Override
    public void onCancel(Object object) {
        eventSink = null;
    }

    @Override
    public View getView() {
        if (lifecycleRegistry.getCurrentState() != Lifecycle.State.RESUMED) {
            lifecycleRegistry.setCurrentState(Lifecycle.State.RESUMED);
        }
        return previewView;
    }

    @NonNull
    @Override
    public Lifecycle getLifecycle() {
        return lifecycleRegistry;
    }


    @Override
    public void dispose() {
        lifecycleRegistry.markState(Lifecycle.State.DESTROYED);
        if (cameraProvider != null) cameraProvider.unbindAll();
    }

    @Override
    public void onListen(Object o, EventChannel.EventSink eventSink) {
        this.eventSink = eventSink;
    }

    @Override
    public void onMethodCall(MethodCall methodCall, MethodChannel.Result result) {
        switch (methodCall.method) {
            case "startScan":
                isPlay = true;
//                result.success(null);x
                break;
            case "stopScan":
                isPlay = false;
//                result.success(null);
                break;
            case "setFlashMode":
                boolean isOpen = methodCall.argument("isOpen");
                cameraControl.enableTorch(isOpen);
                break;
            case "getFlashMode":
                result.success(cameraInfo.getTorchState());
                break;
            default:
                result.notImplemented();
                break;
        }
    }

}

File: android/src/main/kotlin/flutter/curiosity/scan/ScanView.kt
package flutter.curiosity.scan

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.ImageFormat
import android.graphics.Rect
import android.view.View
import android.widget.FrameLayout
import androidx.camera.camera2.Camera2Config
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LifecycleRegistry
import com.google.common.util.concurrent.ListenableFuture
import com.google.zxing.*
import com.google.zxing.common.HybridBinarizer
import flutter.curiosity.CuriosityPlugin.Companion.scanView
import flutter.curiosity.camera.preview.PreviewView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.platform.PlatformView
import java.util.concurrent.Executor
import java.util.concurrent.Executors

@SuppressLint("ViewConstructor")
class ScanView internal constructor(context: Context, messenger: BinaryMessenger, i: Int, any: Any) :
        PlatformView, LifecycleOwner, CameraXConfig.Provider, EventChannel.StreamHandler,
        MethodCallHandler, FrameLayout(context) {
    private lateinit var lifecycleRegistry: LifecycleRegistry
    private lateinit var previewView: PreviewView
    private var isScan: Boolean
    private lateinit var eventSink: EventSink
    private var lastCurrentTime = 0L //最后一次的扫描
    private lateinit var cameraProviderFuture: ListenableFuture<ProcessCameraProvider>
    private lateinit var cameraProvider: ProcessCameraProvider
    private lateinit var cameraControl: CameraControl
    private lateinit var cameraInfo: CameraInfo
    private var multiFormatReader: MultiFormatReader = MultiFormatReader()
    private val executor: Executor = Executors.newSingleThreadExecutor()
    private val anyMap = any as Map<*, *>
    private val topRatio: Int
    private val leftRatio: Int
    private val widthRatio: Int
    private val heightRatio: Int
//    private var mViewFinderView: ViewFinderView? = null

    init {
        isScan = (anyMap["isScan"] as Boolean?)!!
        topRatio = anyMap["topRatio"] as Int
        leftRatio = anyMap["leftRatio"] as Int
        widthRatio = anyMap["widthRatio"] as Int
        heightRatio = anyMap["heightRatio"] as Int
        EventChannel(messenger, "$scanView/$i/event").setStreamHandler(this)
        MethodChannel(messenger, "$scanView/$i/method").setMethodCallHandler(this)
        multiFormatReader.setHints(ScanUtils.hints)
//        initLayout()
        initCameraView()
    }

    private fun initLayout() {
        initCameraView()
//        mViewFinderView = ViewFinderView(context)
//        addView(previewView)
//        addView(mViewFinderView)
    }

    override fun getView(): View {
        if (lifecycleRegistry.currentState != Lifecycle.State.RESUMED) {
            lifecycleRegistry.currentState = Lifecycle.State.RESUMED
        }
        return previewView
    }

    private fun initCameraView() {
        cameraProviderFuture = ProcessCameraProvider.getInstance(context)
        lifecycleRegistry = LifecycleRegistry(this)
        previewView = PreviewView(context)
        val preview = Preview.Builder()
                .setTargetAspectRatio(AspectRatio.RATIO_16_9)
                .build()
        preview.setSurfaceProvider(previewView.previewSurfaceProvider)
        val imageAnalysis = ImageAnalysis.Builder().apply {
            setImageQueueDepth(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
        }.build()
        imageAnalysis.setAnalyzer(executor, ScanImageAnalysis())
        previewView.post { startCamera(context, preview, imageAnalysis) }
    }

    private fun startCamera(context: Context, preview: Preview, imageAnalysis: ImageAnalysis) {
        val cameraSelector = CameraSelector.Builder().requireLensFacing(CameraSelector.LENS_FACING_BACK).build()
        cameraProviderFuture.addListener(Runnable {
            cameraProvider = cameraProviderFuture.get()
            val camera = cameraProvider.bindToLifecycle(this, cameraSelector, preview,
                    imageAnalysis)
            cameraControl = camera.cameraControl
            cameraInfo = camera.cameraInfo
        }, ContextCompat.getMainExecutor(context))
    }

    override fun getCameraXConfig(): CameraXConfig {
        return Camera2Config.defaultConfig()
    }

    override fun onCancel(any: Any) {
    }

    override fun getLifecycle(): Lifecycle {
        return lifecycleRegistry
    }

    override fun dispose() {
        lifecycleRegistry.currentState = Lifecycle.State.DESTROYED
        cameraProvider.unbindAll()
    }

    override fun onListen(o: Any, eventSink: EventSink) {
        this.eventSink = eventSink
    }

    private inner class ScanImageAnalysis : ImageAnalysis.Analyzer {

        override fun analyze(image: ImageProxy) {
            val currentTime = System.currentTimeMillis()
            if (currentTime - lastCurrentTime >= 100L && isScan == java.lang.Boolean.TRUE) {
                if (ImageFormat.YUV_420_888 != image.format) {
                    return
                }
                val rect = image.cropRect
                val buffer = image.planes[0].buffer
                val byteArray = ByteArray(buffer.remaining())
                buffer[byteArray, 0, byteArray.size]
                val source = buildLuminanceSource(byteArray, rect)
                val binaryBitmap = BinaryBitmap(HybridBinarizer(source))
                var result: Result? = null
                try {
                    result = multiFormatReader.decodeWithState(binaryBitmap)
                } catch (e: NotFoundException) {
                    // continue
                } finally {
                    multiFormatReader.reset()
                }
                if (result == null) {
                    val invertedSource = source?.invert()
                    val invertBinaryBitmap = BinaryBitmap(HybridBinarizer(invertedSource))
                    try {
                        result = multiFormatReader.decodeWithState(invertBinaryBitmap)
                    } catch (e: NotFoundException) {
                        // continue
                    } finally {
                        multiFormatReader.reset()
                    }
                }
                if (result != null) resultCode(result)
                buffer.clear()
                lastCurrentTime = currentTime
            }
            image.close()
        }
    }

    fun resultCode(result: Result) {
        previewView.post { eventSink.success(ScanUtils.scanDataToMap(result)) }
    }

    private fun buildLuminanceSource(byteArray: ByteArray, rect: Rect): PlanarYUVLuminanceSource? {
        val newRect = getRectInPreview(rect);
        return PlanarYUVLuminanceSource(byteArray, newRect.width(), newRect.height(), newRect.left, newRect.top,
                newRect.width(), newRect.height(), false)
    }


    override fun onMethodCall(methodCall: MethodCall, result: MethodChannel.Result) {
        when (methodCall.method) {
            "startScan" -> isScan = true
            "stopScan" -> isScan = false
            "setFlashMode" -> {
                val status = methodCall.argument<Boolean>("status")
                if (status != null) {
                    cameraControl.enableTorch(status)
                }
            }
            "getFlashMode" -> result.success(cameraInfo.torchState)
            else -> result.notImplemented()
        }
    }

    @Synchronized
    fun getRectInPreview(rect: Rect): Rect {
        return rect
//        val previewWidth = rect.width()
//        val previewHeight = rect.height()
//        val framingRect = mViewFinderView!!.framingRect
//        val viewFinderViewWidth = mViewFinderView!!.width
//        val viewFinderViewHeight = mViewFinderView!!.height
//        if (framingRect == null || viewFinderViewWidth == 0 || viewFinderViewHeight == 0) {
//            return rect
//        }
//        val newRect = Rect(rect)
//        if (previewWidth < viewFinderViewWidth) {
//            newRect.left = rect.left * previewWidth / viewFinderViewWidth
//            newRect.right = rect.right * previewWidth / viewFinderViewWidth
//        }
//        if (previewHeight < viewFinderViewHeight) {
//            newRect.top = rect.top * previewHeight / viewFinderViewHeight
//            newRect.bottom = rect.bottom * previewHeight / viewFinderViewHeight
//        }
//
//        Utils.logInfo(viewFinderViewWidth.toString())
//        Utils.logInfo(viewFinderViewHeight.toString())
//        Utils.logInfo(newRect.left.toString())
//        Utils.logInfo(newRect.top.toString())
//        Utils.logInfo(newRect.width().toString())
//        Utils.logInfo(newRect.height().toString())
//        return rect
    }

}



